### PR TITLE
Resolved issue #24 on master

### PR DESCRIPTION
**Pull request by NewtonMan**

### CODE REVIEW

File: src/Controller/Component/CacheComponent.php [hunk 1]
```diff
@@ -135,7 +135,10 @@ protected function _writeFile(string $content, string $cache) {
 		if (Configure::read('debug') && !is_dir($folder)) {
 			mkdir($folder, 0770, true);
 		}
-
+        $basefolder = $folder . dirname($cache);
+		if (Configure::read('debug') && !is_dir($basefolder)) {
```
Comment (dereuromark) [inline src/Controller/Component/CacheComponent.php:139]: why is this needed, the true param above also auto creates base folder.
or in other words: we only need this once.

Comment (NewtonMan) [reply src/Controller/Component/CacheComponent.php:139]: I changed the structure to store cached files.

Now the cache directory will have subfolders as show in URL.

For exemple a request to "www.website.com/category/subcategory/my-news-title-goes-here.html" (Cakephp3) will produce a cache file in:
/tmp/cache/views/category/subcategory/my-news-title-goes-here.html

There is another important think, the querystring part (right after ? in URL) will be hashed, for exemple a request to "www.website.com/category/subcategory/my-news-title-goes-here.html?utm_source=somewhere" (Cakephp3) will produce a cache file in:
/tmp/cache/views/category/subcategory/my-news-title-goes-here-fh324j56.html

Working that way there is no problem when you have a huge number of cached files.

**Comment by dereuromark** [comment]
I fixed the CI to GA (travis is dead)
If you rebase upon latest master, we should see the tests for this.

Would also be nice to add 1-2 tests maybe.

**Comment by dereuromark** [comment]
I looked into this and it becomes clear to me that this should not be solved this way

a) the file handling is a deprecated or at least outdated feature, the main way to store the cache is using the engines (which also provide a File one btw).
b) With the engines, you can also use a NestedFile engine extending the File one and provide your own folder nesting etc.

The only thing for now we should fix IMO is the problem of length for this outdated file handling.
The amount of files is a separate concern that should be addressed with the migration to the engine IMO, that would blow up the legacy handling too much.
Especially since you put it into the key generation which is also used for the other engines (that don't need to know about the specific file based issues).
You also only enabled auto path create in debug mode, but for those subfolders this would also be necessary for prod mode.

So to sum things up:
I prepared https://github.com/dereuromark/cakephp-cache/pull/27
- Basic length prevention for legacy files using length check + hash for rest to be < 256
- Still slugging to prevent any special chars or utf8mbX bytes or other dangerous chars to mess up the (file) system per cache.
- Adding an option for custom key generator using config callable, this allows any flexibility in also subfoldering.
